MaxText/layers/attentions.py

```diff
@@ -631,17 +631,28 @@ def apply_attention_dot(
       query = query.astype(jnp.float32)
       key = key.astype(jnp.float32)
 
-    q_seq_len = query.shape[1]
-
     # special sharding for decode
+    q_seq_len = query.shape[1]
+    prefill_qkv_sharding = (BATCH, LENGTH, HEAD, D_KV)
+    decode_qkv_sharding = (DECODE_BATCH, DECODE_LENGTH, HEAD, D_KV)
     if self.is_partition_in_decode(q_seq_len):
-      query = partitioning.with_sharding_constraint(query, (DECODE_BATCH, DECODE_LENGTH, HEAD, D_KV))
-      key = partitioning.with_sharding_constraint(key, (DECODE_BATCH, DECODE_LENGTH, HEAD, D_KV))
-      value = partitioning.with_sharding_constraint(value, (DECODE_BATCH, DECODE_LENGTH, HEAD, D_KV))
+      query = partitioning.with_sharding_constraint(query, decode_qkv_sharding)
+      # avoid sharding scale tensor when using kv cache quantization
+      if self.kv_quant and isinstance(key, KVTensor) and isinstance(value, KVTensor):
+        key.qvalue = partitioning.with_sharding_constraint(key.qvalue, decode_qkv_sharding)
+        value.qvalue = partitioning.with_sharding_constraint(value.qvalue, decode_qkv_sharding)
+      else:
+        key = partitioning.with_sharding_constraint(key, decode_qkv_sharding)
+        value = partitioning.with_sharding_constraint(value, decode_qkv_sharding)
```

Comment:
Nit: a helper method might be helpful for this portion since the decode/prefill blocks are doing almost the same thing